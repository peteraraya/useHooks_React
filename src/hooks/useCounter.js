import { useState } from 'react';

export const useCounter = (initialState = 10) => {

    const [counter, setCounter] = useState(initialState);

    const increment = () => {
        setCounter(counter + 1);
    }

    const decrement = () => {
        setCounter(counter  - 1 );
    }

    const reset = () => {
        setCounter(initialState);
    }

    // exportando
    return {
        counter,
        increment,
        decrement,
        reset,
    }




}




/**
 * InitialState : si no me mandan nada tendremos este valor
 * Este custom hook está extrayendo la logica de mi contador
 * El orden en la destructuraci´´on no importa
 *
*/