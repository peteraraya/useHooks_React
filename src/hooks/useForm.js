import { useState } from "react";

export const useForm = ( initialState = {}) => {

    // menejar el estado de nuestro formulario
    const [values, setValues] = useState(initialState);

    const handleInputChange = ({ target }) => {
        setValues({
            ...values,
            [target.name]: target.value
        });
    }


    return [ values, handleInputChange ];
}



// declaro inisitalState como  {} : para evitar errores

