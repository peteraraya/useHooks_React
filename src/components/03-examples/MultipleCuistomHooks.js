import React from 'react'
import './effects.css';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCuistomHooks = () => {


    const {counter,increment} = useCounter(1);

    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${ counter }`);

    // si trae data extrae la data en la posición 0
    const { author, quote } = !!data && data[0];
   


    return (
        <div>
            <h1>BreakingBas Quotes !!!!!</h1>
            <hr />

            {
                loading  // OPERADOR TERNARIO
                    ?
                    (
                        <div className="alert alert-info text-center">
                            Loading...
                        </div>
                    )
                    :
                    (
                        <blockquote className="blockquote text-right">
                            <p className="mb-0">{ quote }</p>
                    <footer className="blockquote-footer">{ author }</footer>
                        </blockquote>

                    )
            }


            <button
                className="btn btn-primary"
                onClick={ increment }
            > Siguiente frase
            
            </button>


        </div>
    )
}

// con doble negación transformo ese valor en un boleano
// !null = true   |   !!null = false 