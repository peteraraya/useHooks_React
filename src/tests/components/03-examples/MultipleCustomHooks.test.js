import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import '@testing-library/jest-dom';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
// Pasa por un mock : no me interesa la implementación solo los resultados, pero evita llamadas 
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas en <MultipleCustomHooks />', () => {
    
    // l o denimos aqui el useCounter si no va a cambiar

    useCounter.mockReturnValue({
        counter:10,
        incremente: ()=> {}
    })


    test('debe de mostrarse correctamente', () => {

        // data por defecto luego de utilizar mock
        useFetch.mockReturnValue({
            data:null,
            loading:true,
            error:null
        });
        
        const wrapper = shallow(<MultipleCustomHooks /> );

        expect( wrapper ).toMatchSnapshot();
    
    });

    test('debe mostrar la información', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'Pedro',
                quote:'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        // Busquedas en el wrapper
        expect( wrapper.find('.alert').exists()).toBe(false);
        expect( wrapper.find('.mb-0').text().trim()).toBe('Hola Mundo');
        expect( wrapper.find('footer').text().trim()).toBe( 'Pedro' );

        console.log(wrapper.html());
    })
    

})

/**
 * La ventaja de trabajar con un snapshot es que le saca una fotografia actual del componente
 * asi el dia de mañana alguien borra ALGO      
 * Es para decir asi esta mi componenteqiero que se mantenga así
 */
