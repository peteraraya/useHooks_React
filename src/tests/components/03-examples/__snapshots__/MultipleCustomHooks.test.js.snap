// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    BreakingBas Quotes !!!!!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
  >
     Siguiente frase
  </button>
</div>
`;
